lib/src/results.dart

```diff
@@ -207,3 +210,34 @@ extension RealmResultsInternal on RealmResults {
     return RealmResults<T>._(handle, realm);
   }
 }
+
+class RealmResultsIterator<T extends RealmObject> implements Iterator<T> {
+  final Iterable<T> _iterable;
+  final int _length;
+  int _index;
+  T? _current;
+
+  RealmResultsIterator(Iterable<T> iterable)
+      : _iterable = iterable,
+        _length = iterable.length,  
+        _index = 0;
+
+  @override
+  T get current => _current as T;
+
+  @override
+  bool moveNext() {
+    int length = _iterable.length;
+    if (_length != length) {
+      throw ConcurrentModificationError(_iterable);
+    }
+    if (_index >= length) {
+      _current = null;
+      return false;
+    }
+    _current = _iterable.elementAt(_index);
```

Comment:
I see your point. Yes, it could be changed.

Reply:
Come to think of it, I prefer the eager evaluation. It ensures that `calling` current will never throw, despite changes to underlying collection. Instead only `moveNext` will throw. The optimisation for the case where only some of the iterated items are actually accessed is less important in my opinion, so I would keep the `_current` field.

Reply:
Yes, I already returned _current back, also because it has to be evaluated before the index is incremented.

Reply:
I don't have strong opinions on the subject and don't know how the dart runtime does iteration, but if it's anything like the .NET runtime, `moveNext` and `current` will always be called together, so which one throws is kind of irrelevant for the developer. The only minor performance difference is when the enumerator is used to check if the collection is empty, such as `isEmpty = !results.getEnumerator().moveNext()`, but I don't suppose we need to optimize for this.

Reply:
We decided to drop the throw even in `moveNext` (ie. just return false, `if _index >=` current length), so correct iteration should never throw, but it may produce an odd sequence if the underlying collection was changed during the iteration (and the realm result was not snapshotted)

Reply:
Taking in consideration your comments @nielsenko and @nirinchev the code was changed to the following. Snapshotting will be implemented with task #166. Are you OK to resolve this discussion?

`class _RealmResultsIterator<T extends RealmObject> implements Iterator<T> {
  final RealmResults<T> _results;
  int _index;
  T? _current;

  _RealmResultsIterator(RealmResults<T> results)
      : _results = results,
        _index = -1;

  @override
  T get current => _current as T;

  @override
  bool moveNext() {
    int length = _results.length;
    _index++;
    if (_index >= length) {
      _current = null;
      return false;
    }
    _current = _results[_index];

    return true;
  }`